Add validation by schema

Jakeii opened the pull request:
I've added the functionality to be able to validate by passing an object schema to check, checkBody etc. as an alternative to chaining.

example:
````javascript
req.checkBody({
  'email': {
    notEmpty: true,
    isEmail:
      failMsg: 'Invalid Email'
    }
  },
  'password': {
     notEmpty: true
  },
  'name.first': {
    notEmpty: true,
    isLength: {
      options: [2, 10],
      failMsg: 'Must be between 2 and 10 chars long'
    }
  }
});
````
passing true will use the default error message.

rustybailey commented:
Thanks for another great contribution! It looks like the Travis build has failed. Can you check locally and make sure the unit tests are passing?

Jakeii commented:
Ah, I added a last second change as I realised optional wouldn't work. I've gone over it properly now and added a test case for using optional in the schema.

rustybailey commented on lib/express_validator.js:
```diff
@@ -136,6 +136,9 @@ var expressValidator = function(options) {
 
     locations.forEach(function(location) {
       req['check' + _.capitalize(location)] = function(param, failMsg) {
+        if(_.isPlainObject(param)) {
```
I know this is my fault for not including a jscs yet, but let's do a space after the if

rustybailey commented on lib/express_validator.js:
```diff
@@ -153,6 +156,9 @@ var expressValidator = function(options) {
     };
 
     req.check = function(param, failMsg) {
+      if(_.isPlainObject(param)) {
```
space after the if

rustybailey commented on lib/express_validator.js:
```diff
@@ -165,6 +171,29 @@ var expressValidator = function(options) {
+ */
+
+function validateSchema(schema, req, loc, options) {
+  for(var param in schema) {
```
space after the for

rustybailey commented on lib/express_validator.js:
```diff
@@ -165,6 +171,29 @@ var expressValidator = function(options) {
+    loc = (loc === 'any') ? locate(req, param) : loc;
+    var validator = new ValidatorChain(param, null, req, loc, options);
+
+    for(var methodName in schema[param]) {
```
space after the for

rustybailey commented on lib/express_validator.js:
```diff
@@ -165,6 +171,29 @@ var expressValidator = function(options) {
+
+function validateSchema(schema, req, loc, options) {
+  for(var param in schema) {
+    loc = (loc === 'any') ? locate(req, param) : loc;
```
Params not needed around `loc === 'any'`

rustybailey commented on lib/express_validator.js:
```diff
@@ -165,6 +171,29 @@ var expressValidator = function(options) {
+ * validate an object using a schema
+ *
+ * @method validateParams
+ * @param  {Object}       schema    schema of validations
```
I know that this is described in the docs and tests, but it would be nice if the schema was described in detail here, so that it is 100% clear what is expected/allowed.

rustybailey commented on test/checkQuerySchemaTest.js:
```diff
@@ -0,0 +1,124 @@
+    'testparam': {
+      notEmpty: true,
+      isInt: {
+        failMsg: 'Parameter is not an integer'
```
probably just want to reuse `errorMessage` here

rustybailey commented:
Added a few code comments. You also need to add some docs to the README regarding this new functionality.

Jakeii commented:
Sorry I've been too busy to sort this out until now... Sorted those few issues, added some docs.

I'm not sure about `failMsg` though, that is what's used in the code in other places, `errorMessage` is just used in tests.

rustybailey commented:
Thanks for coming back to this!

Good point about `failMsg`. In all of our docs, we refer to things as "errors" with methods like `errorFormatter` and `validationErrors`; we never once use the word "fail". It might make more sense to change `failMsg` to `errorMessage` (just for the schema, not necessarily thoughout the code).

Also, can you add `optional` in the docs about the schema so people know it's available?

Jakeii commented:
Cool, sorted!

rustybailey commented:
:+1:

rustybailey commented:
This is available in 2.16.0
